<p align="left">
  <img src="figures/detectADAM_logo.png" alt="detectADAM logo" width="600"/>
</p>

# **detectADAM**

**detectADAM** (Accession Driven Analysis and Mining) is a Snakemake-based pipeline for the detection and analysis of viral sequences in publicly available sequencing datasets.

---

# **Workflow**

<p align="center">
  <img src="figures/flowchart.png" alt="workflow" width="1000"/>
</p>

Recommended input for the EVE file for masking is the `.fna` output of the pipeline [detectEVE](https://github.com/thackl/detectEVE).

---

# **Quickstart Guide**

```bash
# Download workflow
git clone https://github.com/CelineBos/detectADAM
cd detectADAM

# Install dependencies via mamba or conda (https://github.com/conda-forge/miniforge)
mamba create -n detectADAM
mamba activate detectADAM
mamba env update --file envs/env.yml

# Update config file (if required)
nano config.yaml
##config file is in the below format, please enter necessary paths
accession_file: /pathway/to/accession/txt/file ##REQUIRED
diamond_db: /pathway/to/diamond/database ##REQUIRED
eves: "/pathway/to/eve/file" # file of all EVEs that should be masked from references
reference: "pathway/to/reference/file(s)" # if you have multiple reference, concatenate them into one file

# Run Snakemake
snakemake -c <# of cores>  # e.g. 1 / 8 / all

```

# **Output** 

### In the diamond result folder:

- `<accession>.tsv` diamond blast results from the final contigs generated by the pipeline. 

**NOTE:** The `diamond blast` is running on the assumption the database used also has taxonomic information

`qseqid sseqid stitle pident length mismatch gapopen qstart qend sstart send evalue bitscore sscinames staxids`

Should you not have a diamond database in use with taxonomic information, please remove the `sscinames` & `staxids` from the diamond command

## Retained Files

- retains the final contigs of megahit.

## Index was created

- retains the `.bt2` files for ease of additional use. Delete if you want to change the index.

## Predownloaded FASTQ files

Have your FASTQ files from another source or predownloaded?

Utilize the correct file structure `FASTQ_files/` and put the correct file names in the user accession list, and the pipeline will run as normal, skipping the downloading step.

Ensure that you follow the `_1` and `_2` naming convention for pair end files. 


# **About**

detectADAM was developed as a masters project by C. Bos under the guidance of T. Hackl, S. Lequime & S. van der Meij, at the Rijksuniversiteit Groningen 2025. 

